File: mobile/lib/api/services/support.ts
import { apiClient, type ApiResponse } from "../client";
import { API_ENDPOINTS } from "../config";

export interface NotificationPreferences {
  push_enabled: boolean;
  in_app_enabled: boolean;
  trip_updates: boolean;
  document_updates: boolean;
  readiness_updates: boolean;
  daily_program_updates: boolean;
  support_updates: boolean;
  marketing_updates: boolean;
  support_phone: string;
  support_whatsapp: string;
  support_email: string;
  support_message: string;
  notification_provider_enabled: boolean;
  notification_provider_name: string;
  updated_at: string;
}

export interface DeviceInstallation {
  id: string;
  installation_id: string;
  platform: "IOS" | "ANDROID" | "WEB" | "OTHER";
  device_name: string;
  device_model: string;
  os_version: string;
  app_version: string;
  locale: string;
  timezone: string;
  notifications_enabled: boolean;
  capability_flags: Record<string, unknown>;
  preference_state: Record<string, unknown>;
  provider_token_fields: Record<string, unknown>;
  last_seen_at: string;
  created_at: string;
  updated_at: string;
}

export interface TripSupportUpdate {
  id: string;
  trip_id: string;
  title: string;
  body_md: string;
  urgency: string;
  pinned: boolean;
  publish_at: string;
  package_name: string | null;
  attach_url_signed: string | null;
  created_at: string;
}

export interface DailyProgramItem {
  id: string;
  day_index: number;
  title: string;
  location: string | null;
  notes: string | null;
  start_time: string | null;
  end_time: string | null;
}

export interface DailyProgramDay {
  id: string;
  day_index: number;
  label: string;
  date: string | null;
  items: DailyProgramItem[];
}

export interface DailyProgram {
  trip_id: string;
  trip_code: string;
  trip_name: string;
  trip_status: string;
  current_day_index: number;
  is_trip_live: boolean;
  pinned_resource: {
    id: string;
    title: string;
    description: string;
    resource_type: string;
    viewer_mode: string;
    is_pinned: boolean;
    published_at: string | null;
    file_format: string | null;
    metadata: Record<string, unknown>;
    package_name: string | null;
    file_url_signed: string | null;
  } | null;
  days: DailyProgramDay[];
  updated_at: string | null;
}

export interface TripFeedback {
  id: string;
  booking_reference: string;
  trip_code: string;
  trip_name: string;
  status: "DRAFT" | "SUBMITTED";
  overall_rating: number | null;
  support_rating: number | null;
  accommodation_rating: number | null;
  transport_rating: number | null;
  highlights: string;
  improvements: string;
  testimonial_opt_in: boolean;
  follow_up_requested: boolean;
  review_notes: string;
  reviewed_by_name: string | null;
  reviewed_at: string | null;
  submitted_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface TripFeedbackState {
  eligible: boolean;
  reason: string | null;
  feedback: TripFeedback | null;
}

export interface TripFeedbackPayload {
  status?: "DRAFT" | "SUBMITTED";
  overall_rating?: number;
  support_rating?: number;
  accommodation_rating?: number;
  transport_rating?: number;
  highlights?: string;
  improvements?: string;
  testimonial_opt_in?: boolean;
  follow_up_requested?: boolean;
}

export class SupportService {
  static async getNotificationPreferences(token: string): Promise<ApiResponse<NotificationPreferences>> {
    return apiClient.get<NotificationPreferences>(API_ENDPOINTS.NOTIFICATIONS.PREFERENCES, undefined, token);
  }

  static async updateNotificationPreferences(
    data: Partial<NotificationPreferences>,
    token: string
  ): Promise<ApiResponse<NotificationPreferences>> {
    return apiClient.put<NotificationPreferences>(API_ENDPOINTS.NOTIFICATIONS.PREFERENCES, data, undefined, token);
  }

  static async listDevices(token: string): Promise<ApiResponse<DeviceInstallation[]>> {
    return apiClient.get<DeviceInstallation[]>(API_ENDPOINTS.NOTIFICATIONS.DEVICES, undefined, token);
  }

  static async registerDevice(
    data: Partial<DeviceInstallation> & Pick<DeviceInstallation, "installation_id" | "platform">,
    token: string
  ): Promise<ApiResponse<DeviceInstallation>> {
    return apiClient.post<DeviceInstallation>(API_ENDPOINTS.NOTIFICATIONS.DEVICES, data, undefined, token);
  }

  static async updateDevice(
    deviceId: string,
    data: Partial<DeviceInstallation>,
    token: string
  ): Promise<ApiResponse<DeviceInstallation>> {
    return apiClient.patch<DeviceInstallation>(API_ENDPOINTS.NOTIFICATIONS.DEVICE_DETAIL(deviceId), data, undefined, token);
  }

  static async deleteDevice(deviceId: string, token: string): Promise<ApiResponse<void>> {
    return apiClient.delete<void>(API_ENDPOINTS.NOTIFICATIONS.DEVICE_DETAIL(deviceId), undefined, token);
  }

  static async getTripUpdates(tripId: string, token: string): Promise<ApiResponse<TripSupportUpdate[]>> {
    return apiClient.get<TripSupportUpdate[]>(API_ENDPOINTS.TRIPS.UPDATES(tripId), undefined, token);
  }

  static async getDailyProgram(tripId: string, token: string): Promise<ApiResponse<DailyProgram>> {
    return apiClient.get<DailyProgram>(API_ENDPOINTS.TRIPS.DAILY_PROGRAM(tripId), undefined, token);
  }

  static async getTripFeedback(tripId: string, token: string): Promise<ApiResponse<TripFeedbackState>> {
    return apiClient.get<TripFeedbackState>(API_ENDPOINTS.TRIPS.FEEDBACK(tripId), undefined, token);
  }

  static async saveTripFeedback(
    tripId: string,
    data: TripFeedbackPayload,
    token: string
  ): Promise<ApiResponse<TripFeedback>> {
    return apiClient.post<TripFeedback>(API_ENDPOINTS.TRIPS.FEEDBACK(tripId), data, undefined, token);
  }
}

File: backend/apps/api/serializers/support.py
"""Serializers for Phase 3 mobile pilgrim support endpoints."""

from rest_framework import serializers

from apps.api.serializers.trips import TripResourceSerializer
from apps.common.models import PlatformSettings
from apps.pilgrims.models import DeviceInstallation, NotificationPreference, TripFeedback


class NotificationPreferenceSerializer(serializers.ModelSerializer):
    """Serializer for pilgrim notification preferences."""

    support_phone = serializers.SerializerMethodField()
    support_whatsapp = serializers.SerializerMethodField()
    support_email = serializers.SerializerMethodField()
    support_message = serializers.SerializerMethodField()
    notification_provider_enabled = serializers.SerializerMethodField()
    notification_provider_name = serializers.SerializerMethodField()

    class Meta:
        model = NotificationPreference
        fields = [
            'push_enabled',
            'in_app_enabled',
            'trip_updates',
            'document_updates',
            'readiness_updates',
            'daily_program_updates',
            'support_updates',
            'marketing_updates',
            'support_phone',
            'support_whatsapp',
            'support_email',
            'support_message',
            'notification_provider_enabled',
            'notification_provider_name',
            'updated_at',
        ]
        read_only_fields = [
            'support_phone',
            'support_whatsapp',
            'support_email',
            'support_message',
            'notification_provider_enabled',
            'notification_provider_name',
            'updated_at',
        ]

    def _get_settings(self) -> PlatformSettings:
        return PlatformSettings.get_solo()

    def get_support_phone(self, obj):
        return self._get_settings().mobile_support_phone

    def get_support_whatsapp(self, obj):
        return self._get_settings().mobile_support_whatsapp

    def get_support_email(self, obj):
        return self._get_settings().mobile_support_email

    def get_support_message(self, obj):
        return self._get_settings().mobile_support_message

    def get_notification_provider_enabled(self, obj):
        return self._get_settings().notification_provider_enabled

    def get_notification_provider_name(self, obj):
        return self._get_settings().notification_provider_name


class DeviceInstallationSerializer(serializers.ModelSerializer):
    """Serializer for provider-agnostic device installations."""

    class Meta:
        model = DeviceInstallation
        fields = [
            'id',
            'installation_id',
            'platform',
            'device_name',
            'device_model',
            'os_version',
            'app_version',
            'locale',
            'timezone',
            'notifications_enabled',
            'capability_flags',
            'preference_state',
            'provider_token_fields',
            'last_seen_at',
            'created_at',
            'updated_at',
        ]
        read_only_fields = ['id', 'created_at', 'updated_at', 'last_seen_at']


class TripDailyProgramItemSerializer(serializers.Serializer):
    """Serializer for a single daily-program itinerary item."""

    id = serializers.CharField()
    day_index = serializers.IntegerField()
    title = serializers.CharField()
    location = serializers.CharField(allow_blank=True, allow_null=True)
    notes = serializers.CharField(allow_blank=True, allow_null=True)
    start_time = serializers.DateTimeField(allow_null=True)
    end_time = serializers.DateTimeField(allow_null=True)


class TripDailyProgramDaySerializer(serializers.Serializer):
    """Serializer for a day-grouped daily program."""

    id = serializers.CharField()
    day_index = serializers.IntegerField()
    label = serializers.CharField()
    date = serializers.DateField(allow_null=True)
    items = TripDailyProgramItemSerializer(many=True)


class TripDailyProgramSerializer(serializers.Serializer):
    """Serializer for a booked pilgrim's daily program surface."""

    trip_id = serializers.UUIDField()
    trip_code = serializers.CharField()
    trip_name = serializers.CharField()
    trip_status = serializers.CharField()
    current_day_index = serializers.IntegerField()
    is_trip_live = serializers.BooleanField()
    pinned_resource = TripResourceSerializer(allow_null=True)
    days = TripDailyProgramDaySerializer(many=True)
    updated_at = serializers.DateTimeField(allow_null=True)


class PilgrimTripFeedbackSerializer(serializers.ModelSerializer):
    """Serializer for pilgrim feedback state and submitted content."""

    booking_reference = serializers.CharField(source='booking.reference_number', read_only=True)
    trip_code = serializers.CharField(source='trip.code', read_only=True)
    trip_name = serializers.CharField(source='trip.name', read_only=True)
    reviewed_by_name = serializers.CharField(source='reviewed_by.name', read_only=True)

    class Meta:
        model = TripFeedback
        fields = [
            'id',
            'booking_reference',
            'trip_code',
            'trip_name',
            'status',
            'overall_rating',
            'support_rating',
            'accommodation_rating',
            'transport_rating',
            'highlights',
            'improvements',
            'testimonial_opt_in',
            'follow_up_requested',
            'review_notes',
            'reviewed_by_name',
            'reviewed_at',
            'submitted_at',
            'created_at',
            'updated_at',
        ]
        read_only_fields = [
            'id',
            'booking_reference',
            'trip_code',
            'trip_name',
            'review_notes',
            'reviewed_by_name',
            'reviewed_at',
            'submitted_at',
            'created_at',
            'updated_at',
        ]


class PilgrimTripFeedbackUpsertSerializer(serializers.ModelSerializer):
    """Serializer for draft-save and submission of post-trip feedback."""

    class Meta:
        model = TripFeedback
        fields = [
            'status',
            'overall_rating',
            'support_rating',
            'accommodation_rating',
            'transport_rating',
            'highlights',
            'improvements',
            'testimonial_opt_in',
            'follow_up_requested',
        ]

    def validate(self, attrs):
        """Keep submitted feedback internally consistent."""
        for field_name in ['overall_rating', 'support_rating', 'accommodation_rating', 'transport_rating']:
            value = attrs.get(field_name)
            if value is not None and not 1 <= value <= 5:
                raise serializers.ValidationError({field_name: 'Ratings must be between 1 and 5.'})

        return attrs


class PilgrimTripFeedbackStateSerializer(serializers.Serializer):
    """Serializer for the trip-feedback availability response."""

    eligible = serializers.BooleanField()
    reason = serializers.CharField(allow_blank=True, allow_null=True)
    feedback = PilgrimTripFeedbackSerializer(allow_null=True)
